### ISSUE TITLE
Conflict with traitlets version

### ISSUE THREAD

**Issue by JoyMonteiro** [jupyterhub/jupyterhub#1032]
Hello,

I had posted here a few weeks ago regarding an issue with the PAM authentication in jupyterhub not working ( #996). We solved the issue by upgrading my version of traitlets from 4.1.0 to 4.3.2.

I have two setups of jupyterhub, one with google auth and another with PAM auth (which is working with the above mentioned upgrade). I recently brought down the server with the google auth for maintenance, and after I restarted it, it would no longer spawn single user servers unless I downgraded traitlets to 4.1.0.

My google auth setup is as follows:
* I use sudospawner to run jupyterhub as a regular user.
* Because of the issue that google IDs do not correspond to UNIX users, I used the solution @minrk proposed in another issue  (#853) which subclasses ``LocalProcessSpawner`` to a ``SameUserSpawner`` and logs the user in.

All this works fine with traitlets 4.1.0. When I upgrade traitlets, when sudospawner tries to spawn a single user server (https://github.com/jupyterhub/sudospawner/blob/master/sudospawner/spawner.py#L83)
the ``self`` in ``self.get_env()`` is now ``LocalProcessSpawner`` instead of ``SameUserSpawner``, and an exception is raised when ``get_pwnam(self.user.name)`` is called, as expected  (#853).

How should I go about fixing this? I was not sure whether the issue is with sudospawner or jupyterhub, and so I have posted the issue here. If you think that the issue is with sudospawner, then I will close this one and open another one there.

TIA,
Joy

**Comment by willingc** [comment]
@JoyMonteiro Thanks for the detailed update. We appreciate it :smile:

@minrk is traveling this week so there may be a delayed response. I'll take a look at this today and see if I see anything. Thanks!

**Comment by JoyMonteiro** [comment]
Thanks, @willingc for looking into it :) I'm sure it helps me too if things are explained in detail!

**Comment by minrk** [comment]
@JoyMonteiro can you show the output of `jupyterhub --debug` with traitlets 4.3 and your `jupyterhub_config.py`? I suspect that perhaps something is causing some or all of your config to be ignored.

**Comment by JoyMonteiro** [comment]
@minrk the hub is being used this week by students for their coursework. I think they should be done
within a week. Once that happens, I can bring the server down and try this.

Thanks, and will update soon!

**Comment by willingc** [comment]
Thanks @JoyMonteiro. Please let us know how it goes after the students wrap up their work.

**Comment by JoyMonteiro** [comment]
Hello @willingc @minrk,

Here is the output, as promised. Turns out the Hub is crashing with traitlets 4.3.2 now.

Here is the output on a successful launch using traitlets 4.1.0:

```
[E 2017-04-07 09:31:13.852 JupyterHub app:916] Error adding user 'fff@gmail.com' already in db
    Traceback (most recent call last):
      File "/opt/anaconda3/lib/python3.5/site-packages/jupyterhub/app.py", line 911, in init_users
        yield gen.maybe_future(self.authenticator.add_user(user))
      File "/opt/anaconda3/lib/python3.5/site-packages/jupyterhub/auth.py", line 403, in add_user
        raise KeyError("User %s does not exist." % user.name)
    KeyError: 'User fff@gmail.com does not exist.'
    
{'owner': <SameUserSpawner object at 0x7f5ccbfac358>, 'value': '~', 'trait': <traitlets.traitlets.Unicode object at 0x7f5cc7853c88>}
[D 2017-04-07 09:31:13.874 JupyterHub app:1085] Loading state for xxx@gmail.com from db
[D 2017-04-07 09:31:13.874 JupyterHub app:1097] xxx@gmail.com not running.
{'owner': <SameUserSpawner object at 0x7f5cceec4b38>, 'value': '~', 'trait': <traitlets.traitlets.Unicode object at 0x7f5cc7853c88>}
[D 2017-04-07 09:31:13.879 JupyterHub app:1085] Loading state for yyy@gmail.com from db
[D 2017-04-07 09:31:13.880 JupyterHub app:1097] yyy@gmail.com not running.
{'owner': <SameUserSpawner object at 0x7f5cc7710be0>, 'value': '~', 'trait': <traitlets.traitlets.Unicode object at 0x7f5cc7853c88>}
[D 2017-04-07 09:31:13.885 JupyterHub app:1085] Loading state for zzz@gmail.com from db
[I 2017-04-07 09:31:13.885 JupyterHub app:1089] zzz@gmail.com still running
[D 2017-04-07 09:31:13.887 JupyterHub spawner:582] Polling subprocess every 30s
```
The first error happens all the time, the fff@gmail.com does not exist part. That seems more to do with the Hub trying to find the UNIX user corresponding to the email ID, which does not exist. But the hub runs ok, and functions just fine.

This is the output with traitlets 4.3.2
```
[E 2017-04-07 09:34:44.226 JupyterHub app:916] Error adding user 'fff@gmail.com' already in db
    Traceback (most recent call last):
      File "/opt/anaconda3/lib/python3.5/site-packages/jupyterhub/app.py", line 911, in init_users
        yield gen.maybe_future(self.authenticator.add_user(user))
      File "/opt/anaconda3/lib/python3.5/site-packages/jupyterhub/auth.py", line 403, in add_user
        raise KeyError("User %s does not exist." % user.name)
    KeyError: 'User fff@gmail.com does not exist.'
    
{'trait': <traitlets.traitlets.Unicode object at 0x7faf2b6ab240>, 'value': '~', 'owner': <sudospawner.spawner.SudoSpawner object at 0x7faf2fd23358>}
[D 2017-04-07 09:34:45.693 JupyterHub app:1085] Loading state for xxx@gmail.com from db
[D 2017-04-07 09:34:45.694 JupyterHub app:1097] xxx@gmail.com not running.
{'trait': <traitlets.traitlets.Unicode object at 0x7faf2b6ab240>, 'value': '~', 'owner': <sudospawner.spawner.SudoSpawner object at 0x7faf2fd23be0>}
[D 2017-04-07 09:34:45.804 JupyterHub app:1085] Loading state for yyy@gmail.com from db
[D 2017-04-07 09:34:45.804 JupyterHub app:1097] yyy@gmail.com not running.
{'trait': <traitlets.traitlets.Unicode object at 0x7faf2b6ab240>, 'value': '~', 'owner': <sudospawner.spawner.SudoSpawner object at 0x7faf2fd23f60>}
[D 2017-04-07 09:34:45.946 JupyterHub app:1085] Loading state for zzz@gmail.com from db
sudo: unknown user: zzz@gmail.com
sudo: unable to initialize policy plugin
[E 2017-04-07 09:34:45.956 JupyterHub app:1527]
    Traceback (most recent call last):
      File "/opt/anaconda3/lib/python3.5/site-packages/jupyterhub/app.py", line 1524, in launch_instance_async
        yield self.initialize(argv)
      File "/opt/anaconda3/lib/python3.5/types.py", line 179, in throw
        return self.__wrapped.throw(tp, *rest)
      File "/opt/anaconda3/lib/python3.5/site-packages/jupyterhub/app.py", line 1315, in initialize
        yield self.init_spawners()
      File "/opt/anaconda3/lib/python3.5/site-packages/jupyterhub/app.py", line 1087, in init_spawners
        status = yield spawner.poll()
      File "/opt/anaconda3/lib/python3.5/site-packages/jupyterhub/spawner.py", line 832, in poll
        alive = yield self._signal(0)
      File "/opt/anaconda3/lib/python3.5/site-packages/sudospawner/spawner.py", line 90, in _signal
        reply = yield self.do('kill', pid=self.pid, signal=sig)
      File "/opt/anaconda3/lib/python3.5/site-packages/sudospawner/spawner.py", line 69, in do
        data_str = data_str[data_str.index('{'):data_str.rindex('}')+1]
    ValueError: substring not found

```
Now, it crashes while trying to find a single user server for zzz@gmail.com (which is running, btw, unlike for
{xxx,yyy}@gmail.com).

TIA,
Joy

**Comment by minrk** [comment]
## tl;dr

try adding to `jupyterhub_config.py`:

```python
c.LocalAuthenticator.create_system_users = True
```

which will ensure that a system user exists corresponding to each JupyterHub user.

## details

The `User fff@gmail.com does not exist` error [comes up when](https://github.com/jupyterhub/jupyterhub/blob/0.7.2/jupyterhub/auth.py#L403):

1. a `Local...` Authenticator is in use (e.g. `LocalGoogleOAuthenticator` or PAM)
2. the username doesn't exist on the system (the 'real' unix user, as opposed to the Hub user), and
3. the Authenticator has not been instructed to create system users when users are added to JupyterHub (`c.LocalAuthenticator.create_system_users = False`, the default)

Normally when this is the situation user creation would fail immediately if the user doesn't exist on the system. However, seeing the error on startup means the username has already been added to the Hub's database. My guess is that it is due to a migration from an Authenticator that doesn't need local users (e.g. `GoogleOAuthenticator`) to one that does (e.g. `LocalGoogleOAuthenticator`).

`SameUserSpawner` doesn't need local users to exist, but `SudoSpawner` does. This is an important difference, because sudospawner really does need the usernames to exist, as seen in this error message:

> sudo: unknown user: zzz@gmail.com

If the JupyterHub users should be created on the system, then the answer here is to set the `create_system_users` flag to True in config:

```python
c.LocalAuthenticator.create_system_users = True
```

If the user should already exist and isn't being found, that might be a different issue. Perhaps mapping `@gmail.com` to `zzz`? This can be done in `Authenticator.normalize_username`.

**Comment by JoyMonteiro** [comment]
Thanks, @minrk.

We added the users manually to the config file, hence creating them was not an issue. We really don't want to create system users, since the people using this hub are mostly temporary users.

Does the system user get deleted if the corresponding hub user is deleted? Then, this could be a viable option.

I don't know what traitlets is doing, but if the issue is that `SudoSpawner` required users to exist, then it should have failed with the older versions of traitlets as well. Or am I missing something?

Moreover the two users `xxx@gmail.com` and `yyy@gmail.com` also do not have valid users connected with their IDs. This problem seems to appear **only** if the user has a single user server running. For example, if I shut down all single user servers, and only start one for `xxx@gmail.com`, then the code crashes there.

So, I shut down all single user servers and restarted the Hub with traitlets 4.3.2.
Now, the Hub starts up, and I can authenticate myself. However, when I try to start
a single user server, it gives me a 500 error, with the following traceback. Essentially, we are back to
the `getpwnam()` error.

I can now confirm that I can successfully start a single user server upto traitlets 4.2.2. No version
above that works.

The issue, as you correctly pointed out, is that if traitlets > 4.2.2 is used, `SudoSpawner` looks for a system user, whereas it does not for older releases.

So, is the only fix to set `create_system_users=True`?

This is the traceback, btw.
```
[I 2017-04-07 14:38:50.895 JupyterHub app:1485] JupyterHub is now running at http://127.0.0.1:8000/
[I 2017-04-07 14:39:05.814 JupyterHub log:100] 200 GET /hub/home (xxx@gmail.com@::ffff:80.216.179.110) 55.32ms
[D 2017-04-07 14:39:05.887 JupyterHub log:100] 304 GET /hub/static/css/style.min.css?v=d96e0760e0c2b7356ce89635b646c350 (@::ffff:80.216.179.110) 6.47ms
[D 2017-04-07 14:39:05.892 JupyterHub log:100] 304 GET /hub/static/components/requirejs/require.js?v=6da8be361b9ee26c5e721e76c6d4afce (@::ffff:80.216.179.110) 1.17ms
[D 2017-04-07 14:39:06.092 JupyterHub log:100] 304 GET /hub/logo (@::ffff:80.216.179.110) 1.25ms
[D 2017-04-07 14:39:06.094 JupyterHub log:100] 304 GET /hub/static/js/home.js?v=(& (@::ffff:80.216.179.110) 1.19ms
[D 2017-04-07 14:39:06.184 JupyterHub log:100] 304 GET /favicon.ico (@::ffff:80.216.179.110) 0.98ms
[I 2017-04-07 14:39:08.938 JupyterHub log:100] 302 GET /hub/spawn (xxx@gmail.com@::ffff:80.216.179.110) 9.68ms
[I 2017-04-07 14:39:08.963 JupyterHub log:100] 302 GET /user/xxx@gmail.com (@::ffff:80.216.179.110) 1.65ms
[E 2017-04-07 14:39:09.122 JupyterHub user:251] Unhandled error starting xxx@gmail.com's server: 'getpwnam(): name not found: xxx@gmail.com'
[E 2017-04-07 14:39:09.144 JupyterHub web:1524] Uncaught exception GET /hub/user/xxx@gmail.com (::ffff:80.216.179.110)
    HTTPServerRequest(protocol='http', host='climdyn2.misu.su.se:8000', method='GET', uri='/hub/user/xxx@gmail.com', version='HTTP/1.1', remote_ip='::ffff:80.216.179.110', headers={'Cookie': 'jupyter-hub-token="2|1:0|10:1491568426|17:jupyter-hub-token|44:YWIzYzk2MmE5YjdhNDMzMjg2MjZkMDU3M2U1MzBhN2I=|f8cb7b8dafc03c1ef42804275e2de742dad278c3fbee1da5a77b2924d9391c24"; __utma=91495576.856838739.1474971919.1483041831.1484323824.14; __utmz=91495576.1477770772.7.3.utmcsr=misu.su.se|utmccn=(referral)|utmcmd=referral|utmcct=/about-us/vacancies/phd-student-in-climate-science-1.303281; _ga=GA1.2.856838739.1474971919; s_fid=5586387D3495B1F6-0CBAF45E21E95ED4; __gads=ID=53cae222a52f26ec:T=1475220102:S=ALNI_MZcKN-nMP5hTXn4jSuwU6uVx-zgCQ; is_aaas_user=y; ist_usr=y; __utma=167715490.856838739.1474971919.1484209498.1484853732.9; __utmz=167715490.1484209498.8.4.utmcsr=google|utmccn=(organic)|utmcmd=organic|utmctr=(not%20provided); s_pers=%20c19%3Dsd%253Aproduct%253Ajournal%253Aarticle%7C1485261053952%3B%20v68%3D1485259250112%7C1485261053954%3B%20v8%3D1485259349772%7C1579867349772%3B%20v8_s%3DMore%2520than%25207%2520days%7C1485261149772%3B; _ga=GA1.3.856838739.1474971919', 'User-Agent': 'Mozilla/5.0 (X11; Ubuntu; Linux x86_64; rv:52.0) Gecko/20100101 Firefox/52.0', 'Accept-Encoding': 'gzip, deflate', 'Referer': 'http://climdyn2.misu.su.se:8000/hub/home', 'Accept-Language': 'en-US,en;q=0.5', 'Host': 'climdyn2.misu.su.se:8000', 'Connection': 'close', 'X-Forwarded-Host': 'climdyn2.misu.su.se:8000', 'X-Forwarded-For': '::ffff:80.216.179.110', 'X-Forwarded-Proto': 'http', 'Upgrade-Insecure-Requests': '1', 'X-Forwarded-Port': '8000', 'Accept': 'text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8'})
    Traceback (most recent call last):
      File "/opt/anaconda3/lib/python3.5/site-packages/tornado/web.py", line 1445, in _execute
        result = yield result
      File "/opt/anaconda3/lib/python3.5/site-packages/jupyterhub/handlers/base.py", line 535, in get
        yield self.spawn_single_user(current_user)
      File "/opt/anaconda3/lib/python3.5/site-packages/jupyterhub/handlers/base.py", line 328, in spawn_single_user
        yield gen.with_timeout(timedelta(seconds=self.slow_spawn_timeout), f)
      File "/opt/anaconda3/lib/python3.5/site-packages/jupyterhub/user.py", line 261, in spawn
        raise e
      File "/opt/anaconda3/lib/python3.5/site-packages/jupyterhub/user.py", line 229, in spawn
        ip_port = yield gen.with_timeout(timedelta(seconds=spawner.start_timeout), f)
      File "/opt/anaconda3/lib/python3.5/site-packages/sudospawner/spawner.py", line 83, in start
        reply = yield self.do(action='spawn', args=self.get_args(), env=self.get_env())
      File "/opt/anaconda3/lib/python3.5/site-packages/jupyterhub/spawner.py", line 768, in get_env
        env = self.user_env(env)
      File "/opt/anaconda3/lib/python3.5/site-packages/jupyterhub/spawner.py", line 755, in user_env
        home = pwd.getpwnam(self.user.name).pw_dir
    KeyError: 'getpwnam(): name not found: xxx@gmail.com'
    
[D 2017-04-07 14:39:09.157 JupyterHub base:462] No template for 500
[E 2017-04-07 14:39:09.179 JupyterHub log:99] {
      "Cookie": "jupyter-hub-token=\"2|1:0|10:1491568426|17:jupyter-hub-token|44:YWIzYzk2MmE5YjdhNDMzMjg2MjZkMDU3M2U1MzBhN2I=|f8cb7b8dafc03c1ef42804275e2de742dad278c3fbee1da5a77b2924d9391c24\"; __utma=91495576.856838739.1474971919.1483041831.1484323824.14; __utmz=91495576.1477770772.7.3.utmcsr=misu.su.se|utmccn=(referral)|utmcmd=referral|utmcct=/about-us/vacancies/phd-student-in-climate-science-1.303281; _ga=GA1.2.856838739.1474971919; s_fid=5586387D3495B1F6-0CBAF45E21E95ED4; __gads=ID=53cae222a52f26ec:T=1475220102:S=ALNI_MZcKN-nMP5hTXn4jSuwU6uVx-zgCQ; is_aaas_user=y; ist_usr=y; __utma=167715490.856838739.1474971919.1484209498.1484853732.9; __utmz=167715490.1484209498.8.4.utmcsr=google|utmccn=(organic)|utmcmd=organic|utmctr=(not%20provided); s_pers=%20c19%3Dsd%253Aproduct%253Ajournal%253Aarticle%7C1485261053952%3B%20v68%3D1485259250112%7C1485261053954%3B%20v8%3D1485259349772%7C1579867349772%3B%20v8_s%3DMore%2520than%25207%2520days%7C1485261149772%3B; _ga=GA1.3.856838739.1474971919",
      "User-Agent": "Mozilla/5.0 (X11; Ubuntu; Linux x86_64; rv:52.0) Gecko/20100101 Firefox/52.0",
      "Accept-Encoding": "gzip, deflate",
      "Referer": "http://climdyn2.misu.su.se:8000/hub/home",
      "Accept-Language": "en-US,en;q=0.5",
      "Host": "climdyn2.misu.su.se:8000",
      "Connection": "close",
      "X-Forwarded-Host": "climdyn2.misu.su.se:8000",
      "X-Forwarded-For": "::ffff:80.216.179.110",
      "X-Forwarded-Proto": "http",
      "Upgrade-Insecure-Requests": "1",
      "X-Forwarded-Port": "8000",
      "Accept": "text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8"
    }
[E 2017-04-07 14:39:09.179 JupyterHub log:100] 500 GET /hub/user/xxx@gmail.com (xxx@gmail.com@::ffff:80.216.179.110) 115.89ms
[D 2017-04-07 14:39:09.498 JupyterHub log:100] 304 GET /favicon.ico (@::ffff:80.216.179.110) 1.07ms
```

**Comment by minrk** [comment]
> We really don't want to create system users, since the people using this hub are mostly temporary users.

Then SudoSpawner is probably not the right Spawner for you. SudoSpawner only does one thing: switch to another system user with `sudo` to run the notebook server. This doesn't make sense if you don't want there to be another user to switch to.

> if the issue is that SudoSpawner required users to exist, then it should have failed with the older versions of traitlets as well. Or am I missing something?

You are exactly right. It doesn't make sense that SudoSpawner would ever work with users that don't exist. Traitlets version shouldn't be relevant here, though it could result in slight differences of the error messages. Perhaps another package got upgraded at the same time when traitlets was upgraded?

> The issue, as you correctly pointed out, is that if traitlets > 4.2.2 is used, SudoSpawner looks for a system user, whereas it does not for older releases.

sudo itself looks for the user, which should be true for all versions of sudospawner and traitlets. Do you have logs from sudospawner successfully launching a server as a user that doesn't exist on the system? I would expect the same error you are showing with any version of traitlets.

> So, is the only fix to set create_system_users=True?

If you want to use SudoSpawner, yes, system users must exist. But as you were doing on your other server, using something like SameUserSpawner you can run all your (human) users as the same (system) user. There is a cost: users have permission to access and edit each other's files (and see the config files of JupyterHub itself), because they are all the same user as far as the system is concerned. But if that's okay with you, then this is a simple way to run the whole JupyterHub application as a single user, including the Hub itself.

The main way that I usually run JupyterHub when it's not for users already present on a system is with users in docker containers via DockerSpawner. Containers don't require creating users on the system, so I don't need to manage anything on the system for each user.

**Comment by JoyMonteiro** [comment]
I'm using sudospawner to run the Hub as a lower privilege user, following the docs here:
https://github.com/jupyterhub/jupyterhub/wiki/Using-sudo-to-run-JupyterHub-without-root-privileges

When spawning new single user servers, I use `SameUserSpawner`, which should ideally not call `getpwnam()`, since I override those parts of `LocalProcessSpawner`. I have added some additional
logic to your version of `SameUserSpawner` and done some `chmod`s and `chown`s to provide users with some level of privacy and security. It's not perfect, but it serves our purposes.

Sorry, I think I might have mislead you somehow in the past few posts about what I'm actually doing. My initial post was much clearer. I'm not using `SudoSpawner` to create new single user servers. I'm using `SameUserSpawner`, and I start each single user server as the same user.

So, this worked just fine with all versions of traitlets until 4.2.2. From 4.3.0 onwards, my `SameUserSpawner` is no longer being used, and for some reason `LocalProcessSpawner` is being called, which is calling  `getpwnam()` and causing things to fail.

**Comment by minrk** [comment]
@JoyMonteiro JupyterHub only uses one Spawner, so you can't be using both SudoSpawner and SameUserSpawner at the same time. Your logs above show that sometimes one case using SudoSpawner and another using SameUserSpawner. If you are using SameUserSpawner, you also don't need elevated permissions.

Can you share your `jupyterhub_config.py` and your complete log from starting `jupyterhub --debug`?

**Comment by JoyMonteiro** [comment]
@minrk I'm attaching the output from three versions, and the config.
I have scrubbed them a bit, but none of the configuration information.

I also tried your suggestion, and did not pass the `SudoSpawner` argument during invocation.
I used to
```bash
sudo -u jupyterhub jupyterhub --JupyterHub.spawner_class=sudospawner.SudoSpawner
```
This actually works! The server needs to run for a few days and people have to use it to verify this statement, but I'm able to start single user servers without problems.

If you see the logs, particularly log4.1.0.txt, at the very beginning, the output says that the server recognises that the config has changed, and changes the spawner accordingly.

The 4.2.2 output is not as verbose, but I believe it is also doing this correctly.

In 4.3.2, it does not recognise the change in the spawner from SudoSpawner specified during invocation to SameUserSpawner when reading the config file.

I don't know if this is expected behaviour or a bug, but I think that this is the difference between the scenario where things work and they don't.

[log4.1.0.txt](https://github.com/jupyterhub/jupyterhub/files/907353/log4.1.0.txt)
[log4.3.2.txt](https://github.com/jupyterhub/jupyterhub/files/907354/log4.3.2.txt)
[log4.2.2.txt](https://github.com/jupyterhub/jupyterhub/files/907356/log4.2.2.txt)
[jupyterhub_config.py.txt](https://github.com/jupyterhub/jupyterhub/files/907357/jupyterhub_config.py.txt)

**Comment by minrk** [comment]
Since JupyterHub only uses one Spawner, this line:

```
sudo -u jupyterhub jupyterhub --JupyterHub.spawner_class=sudospawner.SudoSpawner
```

should result in using SudoSpawner and ignoring the SameUserSpawner config in your `jupyterhub_config.py`. There was [a bug](https://github.com/ipython/traitlets/pull/250) in traitlets 4.0-4.2, where the reverse would happen: the config file would override options on the command-line. So in traitlets 4.2, the `--JupyterHub.spawner_class=...` arg was doing nothing (a fortunate coincidence, because you want to be using SameUserSpawner here, not SudoSpawner). traitlets 4.3 fixed the bug where the erroneous argument was being ignored, causing SudoSpawner to be used instead of SameUserSpawner, as requested. Removing the SudoSpawner arg should get you back to using SameUserSpawner, like before.

Some points about your config:

I see that you've selected `LocalGoogleOAuthenticator`. You only want to use the `Local` variant if each Google account should have a corresponding account on the system. Since you are using `SameUserSpawner`, this is not the case. Instead, you probably want to use the base `GoogleOAuthenticator`, which has no relationship to local users. This will fix the "User xxx does not exist" messages on startup.

I see that your SameUserSpawner is calling `setuid` like the default LocalProcessSpawner, apparently to switch to the `jupyterhub` user. `setuid` can only be called successfully in two cases:

1. current user is root and it changes to the target user
2. current user and target are the same, where it does nothing

The bulk of `SameUserSpawner` is "don't switch users", which is why the [original implementation](https://github.com/jupyterhub/jupyterhub/issues/853#issuecomment-271888795) made `preexec_fn` a no-op. The `try: setuid()...` version in your config file doesn't accomplish much that I can see, and might be better replaced with a no-op.

**Comment by JoyMonteiro** [comment]
Things make much more sense now!!

Thanks for the pointers for my config, it does indeed get rid of those warnings, and cleared some
misconceptions about how things work.
